### DISCUSSION TITLE
How to control the icubSim speed in position direct mode

### DISCUSSION THREAD

**Discussion by FranciscoNaveros** [robotology/community#251] [Q&A] [answered]
Hi,
I'm trying to develop an horizontal VOR experiment using the iCub simulator (yarp-2.3.70 and icub-main-1.8.0). The head must be able to describe a sinusoidal trajectory and the eyes must compensate this movement. 

I'm using the position direct mode, sending a position command each 2 ms. The problem it is that the head and eyes movements are too slow and the simulated robot is not able to describe the desired trajectory. The resulting trajectory is a bit smaller and quite delayed (more than 50ms) respect to the desired one. In the normal position mode I can control the speed and aceleration using the functions setRefSpeeds and setRefAccelerations respectively. How could I increase the speed in the position direct mode?. I have tried to modify the PID parameters of second and fourth head joints (horizontal head and eyes movements), but the speed is the same. 

![figure](https://user-images.githubusercontent.com/14795153/30652014-88007896-9e27-11e7-849f-fdbeb402d8f0.jpg)

Thanks in advance, 

Francisco Naveros.

**Comment by pattacini** [comment, marked as answer]
Hi @FranciscoNaveros and welcome to Robotology.

The `position-direct` mode is very peculiar in that it lets you gain direct access to the PID setpoints at the low-level. Therefore, you'll bypass straightaway the trajectory generator that allows you to tinker with the speed and acceleration settings, which are by contrast available only for the standard `position` mode (not suitable for your goal).

For deeper insights on the differences among control modes, you may consider exploring this [**walk-through**][1].

In brief, all that matters is about fine-tuning the PID gains, as you've already figured out yourself. This process might not be that easy though, requiring some investigations of the joint model implemented by the simulator and, often, a principled yet time-consuming approach (i.e. system identification along with control design techniques) gives significant advantages over attempts based on trial and error.

In addition, controlling at a rate of `500 Hz` could be problematic depending on the resources you have available. I reckon such a high rate is not strictly necessary for your experiments. Personally, I'd start off playing with `100 Hz`, by which we had quite a successful gaze control thus far.

Providing a [**Gist**][2] with a very minimal code snippet that implements your experiment could be of help, as per our [support guidelines][3].

[1]: https://github.com/vvv-school/tutorial_joint-interface
[2]: https://gist.github.com
[3]: https://github.com/robotology/QA/blob/master/.github/SUPPORT.md
